crates/iceberg/src/spec/table_metadata_builder.rs

```diff
@@ -0,0 +1,2070 @@
+                removed_snapshots.push(*k);
+                false
+            } else {
+                true
+            }
+        });
+
+        if !removed_snapshots.is_empty() {
+            self.changes.push(TableUpdate::RemoveSnapshots {
+                snapshot_ids: removed_snapshots,
+            });
+        }
+
+        // Remove refs that are no longer valid
+        self.metadata
+            .refs
+            .retain(|_, v| self.metadata.snapshots.contains_key(&v.snapshot_id));
+
+        self
+    }
+
+    /// Set a reference to a snapshot.
+    ///
+    /// # Errors
+    /// - The snapshot id is unknown.
+    pub fn set_ref(mut self, ref_name: &str, reference: SnapshotReference) -> Result<Self> {
+        if self
+            .metadata
+            .refs
+            .get(ref_name)
+            .is_some_and(|snap_ref| snap_ref.eq(&reference))
+        {
+            return Ok(self);
+        }
+
+        let Some(snapshot) = self.metadata.snapshots.get(&reference.snapshot_id) else {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                format!(
+                    "Cannot set '{ref_name}' to unknown snapshot: '{}'",
+                    reference.snapshot_id
+                ),
+            ));
+        };
+
+        // Update last_updated_ms to the exact timestamp of the snapshot if it was added in this commit
+        let is_added_snapshot = self.changes.iter().any(|update| {
+            matches!(update, TableUpdate::AddSnapshot { snapshot: snap } if snap.snapshot_id() == snapshot.snapshot_id())
+        });
+        if is_added_snapshot {
+            self.metadata.last_updated_ms = snapshot.timestamp_ms();
+        }
+
+        // Current snapshot id is set only for the main branch
+        if ref_name == MAIN_BRANCH {
+            self.metadata.current_snapshot_id = Some(snapshot.snapshot_id());
+            if self.metadata.last_updated_ms == i64::default() {
+                self.metadata.last_updated_ms = chrono::Utc::now().timestamp_millis();
+            };
+
+            self.metadata.snapshot_log.push(SnapshotLog {
+                snapshot_id: snapshot.snapshot_id(),
+                timestamp_ms: self.metadata.last_updated_ms,
+            });
+        }
+
+        self.changes.push(TableUpdate::SetSnapshotRef {
+            ref_name: ref_name.to_string(),
+            reference: reference.clone(),
+        });
+        self.metadata.refs.insert(ref_name.to_string(), reference);
+
+        Ok(self)
+    }
+
+    /// Remove a reference
+    ///
+    /// If `ref_name='main'` the current snapshot id is set to -1.
+    pub fn remove_ref(mut self, ref_name: &str) -> Self {
+        if ref_name == MAIN_BRANCH {
+            self.metadata.current_snapshot_id = Some(i64::from(Self::LAST_ADDED));
+            self.metadata.snapshot_log.clear();
+        }
+
+        if self.metadata.refs.remove(ref_name).is_some() || ref_name == MAIN_BRANCH {
```

Comment:
I think we should keep it actually.

A few lines above we have `if ref_name == MAIN_BRANCH {}`. Thus, if `ref_name` is main, then we mutate the internal state of the builder - even if the branch does not exist. Because we mutate the state of the builder, we should record the change.

So I think we have two choices:
1. Leave the `|| ref_name == MAIN_BRANCH`
2. Only clear snapshot log & set last snapshot to None if main branch actually exists

@Fokko I guess its a nit, but Java has the same problem:
If main does not exist the internal state of the builder is mutated but the change not recorded in changes:
https://github.com/apache/iceberg/blob/daa24f9c3a56e18d188097deb2dd79cc991c9a78/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1269-L1271

Reply:
Both ours and the java builder handle removing the main ref - in rust right above your comment. In java here:
https://github.com/apache/iceberg/blob/799925a4ef41e7b4231930377b83bd686001c2c0/core/src/main/java/org/apache/iceberg/TableMetadata.java#L1264

However I couldn't delete it with spark. So I dug deeper. Its prohibited here:
https://github.com/apache/iceberg/blob/799925a4ef41e7b4231930377b83bd686001c2c0/core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java#L71

The spec states:
"There is always a main branch reference pointing to the current-snapshot-id even if the refs map is null."

I think the spec is not very precise, as a table without snapshots can't have a `current-snapshot-id` - the field is nullable in the spec for a reason. I interpret the spec intention as: It should not be removed once it's there.

Of course we could have two consecutive updates: Delete main ref followed by set main ref. Not sure if any clients actually produce this - it would be a way to "force-push" the main branch, truncating all its history.
To stay spec compliant I added a `ensure_main_branch_not_removed` method and call it during `build` and kept the `|| ref_name == MAIN_BRANCH` to cover the "remove & set in one transaction" case.

Commit: https://github.com/apache/iceberg-rust/pull/587/commits/e4dcc7fc094c5fbe4e50bc6e8e9435ad02d63170